File: pipeline/detectors/python_name_error.py
"""
Detector for Python NameError exceptions.
"""

import re
import typing as T
from pipeline.detectors.base import Detector
from pipeline.models import ErrorClue


class PythonNameErrorDetector(Detector):
    """
    Detect Python NameError exceptions indicating missing imports or code.

    Matches patterns like:
    - NameError: name 'fcntl' is not defined
    - NameError: global name 'SomeClass' is not defined
    """

    PATTERNS = {
        "python_name_error": r'File "(?P<file_path>[^"]+\.py)", line (?P<line_number>\d+),.*?NameError: (?:global )?name \'(?P<undefined_name>\w+)\' is not defined',
    }

    EXAMPLES = [
        (
            'File "test.py", line 5, in test_func\nNameError: name \'fcntl\' is not defined',
            {
                "clue_type": "python_name_error",
                "confidence": 1.0,
                "context": {
                    "file_path": "test.py",
                    "undefined_name": "fcntl",
                    "line_number": "5",
                },
            },
        ),
    ]

File: src/boil/detectors/python_name_error.go
// Detector for Python NameError exceptions.
package detectors

// PythonNameErrorDetector detects Python NameError exceptions
type PythonNameErrorDetector struct {
	*BaseDetector
}

func NewPythonNameErrorDetector() (*PythonNameErrorDetector, error) {
	base, err := NewBaseDetector(
		"PythonNameErrorDetector",
		100,
		map[string]string{
			"python_name_error": `File "(?P<file_path>[^"]+\.py)", line (?P<line_number>\d+),.*?NameError: (?:global )?name '(?P<undefined_name>\w+)' is not defined`,
		},
		[]DetectorExample{
			{
				Name:     "python_name_error",
				Input:    `File "test.py", line 10, in main NameError: name 'undefined_var' is not defined`,
				ClueType: "python_name_error",
				Context: map[string]string{
					"file_path":      "test.py",
					"line_number":    "10",
					"undefined_name": "undefined_var",
				},
			},
		},
	)
	if err != nil {
		return nil, err
	}
	return &PythonNameErrorDetector{base}, nil
}
